### Levene's Test 是什么？

Levene 检验用于检验为两个或更多组（样本）计算的变量方差是否相等。


### 如何解释Levene's Test  ？

Levene检验的零假设是，所有组的方差相等

Levene检验的备择假设是，至少有一对组的方差不相等

如果Levene检验的p值小于显著性水平（例如0.05），则至少两组的方差不相等。

In [1]:
import pandas as pd
from scipy.stats import levene

In [3]:
df = pd.read_csv('data_test.csv')
df.head(50)

group  before_treatment  after_treatment
0      control              27.9             33.8
1      control              16.8              9.3
2      control              27.2             23.4
3      control              12.5             19.9
4      control              14.4             16.0
5      control              13.5             19.9
6      control              17.1             21.1
7      control              21.4             16.9
8      control              18.5             17.8
9      control              18.3             15.2
10     control              20.8             18.6
11     control              19.6             21.2
12     control              17.1             14.8
13     control              25.4             24.9
14     control              17.3             15.1
15     control              21.9             23.9
16     control              17.5             25.8
17     control              19.6             14.9
18     control              22.6             19.1
19     control              19.5             15.8
20     control              19.5             21.3
21     control              17.4             10.1
22     control              17.6             12.3
23     control              25.1             22.4
24     control              22.2             23.6
25     control              28.0             34.0
26     control              23.2             23.6
27     control              20.6             19.4
28     control              14.2             19.2
29     control              19.5             22.2
30     control              27.0             22.5
31     control              28.2             24.7
32     control              21.9             27.8
33     control              15.0             17.6
34     control              20.7             11.2
35     control              23.7             18.5
36     control              25.0             18.7
37     control              12.9             16.9
38     control              15.8             18.4
39     control              19.4             21.2
40  treatment1              19.5             15.5
41  treatment1              16.9             12.3
42  treatment1              22.9             14.0
43  treatment1              25.3             17.8
44  treatment1              19.9             15.7
45  treatment1              25.2             16.3
46  treatment1              12.7             15.2
47  treatment1              21.9             16.1
48  treatment1              20.2             18.5
49  treatment1              13.3             11.4

In [4]:
df_agg = (
    df.groupby("group")
    .agg(
        avg_bef_treat=("before_treatment", "mean"),
        var_bef_treat=("before_treatment", "var"),
        avg_aft_treat=("after_treatment", "mean"),
        var_aft_treat=("after_treatment", "var"),
    )
    .reset_index()
)

print(df_agg)

        group  avg_bef_treat  var_bef_treat  avg_aft_treat  var_aft_treat
0     control         20.145      18.878436         19.825      28.825513
1  treatment1         19.210      17.007263         15.475       4.649342
2  treatment2         21.510      19.673579         20.315      15.141458


Here you can clearly see the difference between “var_bef_tr” (variance before treatment) across 3 groups is not that different: 18.88, 17.01, 19.67.

However the difference between “var_aft_tr” (variance after treatment) across 3 groups is quite different: 28.83, 4.65, 15.14.

The difference in variance after treatment across three groups is large enough for us to be almost sure that it is significantly different, but to check it statistically we will perform the Levene’s test in Python!

We will need to create the variables which will store observations relevant to a particular group:

In [5]:
control_group = df[df['group']=='control']['after_treatment']

treatment1_group = df[df['group']=='treatment1']['after_treatment']

treatment2_group = df[df['group']=='treatment2']['after_treatment']

Finally perform the Levene’s test in Python:

In [6]:
stat, p_value = levene(control_group, treatment1_group, treatment2_group)

print(f"Lavene's test statistic: {stat}")
print(f"P-value: {p_value}")

Lavene's test statistic: 4.240957881271611
P-value: 0.017895111992486838


由于p值小于0.05，我们拒绝了零假设，并得出结论，至少有一对组具有不等方差（不是我们想要的“方差齐性”数据）

## 作业附加题（5分 bonus）：

### 在该数据里，请对每组数据（control, treatment1, treatment2, + before treatment, after treatment）做正态检验。

### 在该数据里，对不同的组别间进行比较，找出哪些组群的方差是一样的。